core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -0,0 +1,264 @@
+    false
+  }
+
+  private def aclMatch(session: Session, operations: Operation, resource: Resource, principal: KafkaPrincipal, host: String, permissionType: PermissionType, acls: Set[Acl]): Boolean = {
+    for (acl: Acl <- acls) {
+      if (acl.permissionType.equals(permissionType)
+        && (acl.principal.equals(principal) || acl.principal.equals(Acl.WildCardPrincipal))
+        && (operations.equals(acl.operation) || acl.operation.equals(All))
+        && (acl.host.contains(host) || acl.host.contains(Acl.WildCardHost))) {
+        debug(s"operation = $operations on resource = $resource from host = $host is $permissionType.name based on acl = $acl")
+        return true
+      }
+    }
+    false
+  }
+
+  def syncAcls(): Unit = {
+    debug("Syncing the acl cache for all resources that are currently in cache ")
+    var resources: collection.Set[Resource] = Set.empty
+    inReadLock(lock) {
+      resources = aclCache.keySet
+    }
+
+    for (resource: Resource <- resources) {
+      val resourceAcls: Set[Acl] = getAcls(resource)
+      inWriteLock(lock) {
+        aclCache.put(resource, resourceAcls)
+      }
+    }
+  }
+
+  override def addAcls(acls: Set[Acl], resource: Resource): Unit = {
+    if (acls == null || acls.isEmpty)
+      return
+
+    val updatedAcls: Set[Acl] = getAcls(resource) ++ acls
+    val path: String = toResourcePath(resource)
+
+    if (ZkUtils.pathExists(zkClient, path))
+      ZkUtils.updatePersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(updatedAcls)))
+    else
+      ZkUtils.createPersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(updatedAcls)))
+
+    inWriteLock(lock) {
+      aclCache.put(resource, updatedAcls)
+    }
+
+    updateAclChangedFlag(resource)
+  }
+
+  override def removeAcls(acls: Set[Acl], resource: Resource): Boolean = {
+    if (!ZkUtils.pathExists(zkClient, toResourcePath(resource)))
+      return false
+
+    val existingAcls: Set[Acl] = getAcls(resource)
+    val filteredAcls: Set[Acl] = existingAcls.filter((acl: Acl) => !acls.contains(acl))
+    if (existingAcls.equals(filteredAcls))
+      return false
+
+    val path: String = toResourcePath(resource)
+    ZkUtils.updatePersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(filteredAcls)))
+
+    inWriteLock(lock) {
+      aclCache.put(resource, filteredAcls)
+    }
+
+    updateAclChangedFlag(resource)
+
+    true
+  }
+
+  override def removeAcls(resource: Resource): Boolean = {
+    if (ZkUtils.pathExists(zkClient, toResourcePath(resource))) {
+      ZkUtils.deletePath(zkClient, toResourcePath(resource))
+      inWriteLock(lock) {
+        aclCache.remove(resource)
+      }
+      updateAclChangedFlag(resource)
+      true
+    } else false
+  }
+
+  override def getAcls(resource: Resource): Set[Acl] = {
+    inReadLock(lock) {
+      if (aclCache.contains(resource))
+        return aclCache.get(resource).get
+    }
```

Comment:
It seems that it will be cleaner if we split this into getAcls and getAclsFromZK. The former will just read from the cache and the latter always read from ZK. Only the ZK listener will use the latter.